go/consensus/tendermint/apps/beacon/slashing.go

```diff
@@ -0,0 +1,112 @@
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+)
+
+func onSCRAPEMisbehavior(
+	ctx *abciAPI.Context,
+	id signature.PublicKey,
+	why staking.SlashReason,
+) error {
+	// Retrieve the penalty for the slashing reason.
+	stakeState := stakingState.NewMutableState(ctx.State())
+	st, err := stakeState.Slashing(ctx)
+	if err != nil {
+		ctx.Logger().Error("failed to get slashing table",
+			"err", err,
+		)
+		return fmt.Errorf("beacon: failed to get slashing table: %w", err)
+	}
+	penalty := st[why]
+
+	// Retrieve the node descriptor.
+	regState := registryState.NewMutableState(ctx.State())
+	node, err := regState.Node(ctx, id)
+	if err != nil {
+		// This should never happen since transactions only come from
+		// nodes that are registered and selected as participants.
+		ctx.Logger().Error("failed to get node by signature public key",
+			"public_key", id,
+			"err", err,
+		)
+		return fmt.Errorf("beacon: failed to get node by id %s: %w", id, err)
+	}
+
+	// Freeze the node if the penalty for this class of misbehavior
+	// includes a non-zero freeze time.
+	if penalty.FreezeInterval > 0 {
+		var nodeStatus *registry.NodeStatus
+		if nodeStatus, err = regState.NodeStatus(ctx, node.ID); err != nil {
+			// This should never happen either, since all validators
+			// should have valid node statuses.
+			ctx.Logger().Warn("failed to get validator node status",
+				"err", err,
+				"node_id", node.ID,
+			)
+			return fmt.Errorf("beacon: failed to get node status: %w", err)
+		}
+
+		// If the node is already frozen, don't slash it again.
+		//
+		// XXX: Nodes that repeatedly attempt to misbehave *should* be
+		// slashed into oblivion, but people will probably cry about it.
+		if nodeStatus.FreezeEndTime > 0 {
+			ctx.Logger().Warn("refusing to slash a frozen node",
+				"node_id", node.ID,
+			)
+			return nil
+		}
+
+		var epoch beacon.EpochTime
+		epoch, err = ctx.AppState().GetEpoch(context.Background(), ctx.BlockHeight()+1)
+		if err != nil {
+			return err
+		}
+
+		// Check for overflow.
+		if math.MaxUint64-penalty.FreezeInterval < epoch {
+			nodeStatus.FreezeEndTime = registry.FreezeForever
+		} else {
+			nodeStatus.FreezeEndTime = epoch + penalty.FreezeInterval
+		}
+
+		if err = regState.SetNodeStatus(ctx, node.ID, nodeStatus); err != nil {
+			ctx.Logger().Error("failed to set validator node status",
+				"err", err,
+				"node_id", node.ID,
+				"entity_id", node.EntityID,
+			)
+			return fmt.Errorf("beacon: failed to update node status: %w", err)
+		}
+	}
+
+	// Slash runtime node entity.
+	entityAddr := staking.NewAddress(node.EntityID)
+	totalSlashed, err := stakeState.SlashEscrow(ctx, entityAddr, &penalty.Amount)
+	if err != nil {
+		return fmt.Errorf("beacon: error slashing account %s: %w", entityAddr, err)
+	}
+	if totalSlashed.IsZero() && totalSlashed.Cmp(&penalty.Amount) != 0 {
+		return fmt.Errorf("beacon: nothing to slash from account %s", entityAddr)
```

Comment:
The second condition probably shouldn't trigger an error as the penalty could be higher than the minimum amount required for a validator. Otherwise this can likely be used as a DoS vector under a specific stake distribution.

I assume the first condition (non-zero slashing) is fine since only active validators can be slashed and those need at least the minimum amount of stake (either in escrow or debonding). But in case a network would be configured with a zero threshold for validators, this could be used to trigger a chain halt.

Reply:
Well, no because the amount check boils down to "AND 0 == penalty.Amount", but I removed it anyway to support more exotic configurations.